File: Recursion and Backtracking/python/nQueens.py
# Problem Statement: The n-queens is the problem of placing n queens on n x n chessboard such that no two queens can attack each other. 
# Given an integer n, return all distinct solutions to the n -queens puzzle. Each solution contains a distinct boards configuration of the queen's placement, where 'Q' and '.' indicate queen and empty space respectively.

# Problem Link: https://www.codingninjas.com/studio/problems/n-queens_696453?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

def construct(board):
    res = []
    for row in board:
        for element in row:
            res.append(element)
    return res

def solve(col, board, ans, left_row, upper_diagonal, lower_diagonal, n):
    if col == n:
        ans.append(construct(board))
        return

    for row in range(n):
        if left_row[row] == 0 and upper_diagonal[row + col] == 0 and lower_diagonal[n - 1 + col - row] == 0:
            board[row][col] = 1
            left_row[row] = 1
            upper_diagonal[row + col] = 1
            lower_diagonal[n - 1 + col - row] = 1

            solve(col + 1, board, ans, left_row, upper_diagonal, lower_diagonal, n)

            board[row][col] = 0
            left_row[row] = 0
            upper_diagonal[row + col] = 0
            lower_diagonal[n - 1 + col - row] = 0

def n_queens(n):
    board = [[0] * n for _ in range(n)]
    ans = []
    left_row = [0] * n
    upper_diagonal = [0] * (2 * n - 1)
    lower_diagonal = [0] * (2 * n - 1)

    solve(0, board, ans, left_row, upper_diagonal, lower_diagonal, n)
    return ans

File: Recursion and Backtracking/cpp/nQueens.cpp
// Problem Statement: The n-queens is the problem of placing n queens on n x n chessboard such that no two queens can attack each other. 
// Given an integer n, return all distinct solutions to the n -queens puzzle. Each solution contains a distinct boards configuration of the queen's placement, where 'Q' and '.' indicate queen and empty space respectively.

// Problem Link: https://www.codingninjas.com/studio/problems/n-queens_696453?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
#include <vector>
using namespace std;

vector<int> construct(vector<vector<int>>& board) {
    vector<int> res;
    for (const auto& row : board) {
        for (int element : row) {
            res.push_back(element);
        }
    }
    return res;
}

void solve(int col, vector<vector<int>>& board, vector<vector<int>>& ans, 
           vector<int>& leftRow, vector<int>& upperDiagonal, vector<int>& lowerDiagonal, int n) {
    if (col == n) {
        ans.push_back(construct(board));
        return;
    }

    for (int row = 0; row < n; row++) {
        if (leftRow[row] == 0 && upperDiagonal[row + col] == 0 && lowerDiagonal[n - 1 + col - row] == 0) {
            board[row][col] = 1;
            leftRow[row] = 1;
            upperDiagonal[row + col] = 1;
            lowerDiagonal[n - 1 + col - row] = 1;

            solve(col + 1, board, ans, leftRow, upperDiagonal, lowerDiagonal, n);

            board[row][col] = 0;
            leftRow[row] = 0;
            upperDiagonal[row + col] = 0;
            lowerDiagonal[n - 1 + col - row] = 0;
        }
    }
}

vector<vector<int>> nQueens(int n) {
    vector<vector<int>> board(n, vector<int>(n, 0));
    vector<vector<int>> ans;
    vector<int> leftRow(n, 0);
    vector<int> upperDiagonal(2 * n - 1, 0);
    vector<int> lowerDiagonal(2 * n - 1, 0);

    solve(0, board, ans, leftRow, upperDiagonal, lowerDiagonal, n);
    return ans;
}
